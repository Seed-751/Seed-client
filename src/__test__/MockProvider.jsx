import React from "react";
import PropTypes from "prop-types";
import { Provider } from "react-redux";
import { ThemeProvider } from "styled-components";
import GrobalStyles from "../styles/index";
import theme from "../styles/theme";
import store from "../reducers/store";
import { Router } from "react-router-dom";
import { createMemoryHistory } from "history";

export default function MockProvider({ children }) {
  const history = createMemoryHistory();

  return (
    <Provider store={store}>
      <Router history={history}>
        <ThemeProvider theme={theme}>
          <GrobalStyles />
          {children}
        </ThemeProvider>
      </Router>
    </Provider>
  );
}

MockProvider.propTypes = {
  children: PropTypes.element.isRequired,
};
