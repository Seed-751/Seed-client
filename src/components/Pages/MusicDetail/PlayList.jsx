import React from "react";
import { useDispatch } from "react-redux";
import PropTypes from "prop-types";
import styled from "styled-components";

import { IconButton } from "@material-ui/core/";
import { PlayArrow } from "@material-ui/icons/";
import { setCurrentTrack } from "../../../reducers/currentMusicSlice";

const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  gap: 10px;
  width: 100%;

  strong {
    font-size: 25px;
  }
`;

const TrackBox = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  width: 100%;
  border-bottom: 1px solid ${({ theme }) => theme.color.lightGray};
  line-height: 60px;
  color: ${({ theme }) => theme.color.lightGray};

  .title {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    width: 40%;
  }

  .icon {
    color: white;
    font-size: 30px;
  }
`;

export default function PlayList({ musics }) {
  const dispatch = useDispatch();

  function handleSetCurrentTrack(music) {
    dispatch(setCurrentTrack(music));
  }

  return (
    <Wrapper>
      <strong>Tracks</strong>
      {musics?.map((music) => {
        const { title, track, _id } = music;
        const currentTrack = track.split("/")[0];
        return (
          <TrackBox key={_id}>
            <p>{currentTrack}</p>
            <div className="title">
              <p>{title}</p>
            </div>
            <IconButton onClick={() => handleSetCurrentTrack(music)}>
              <PlayArrow className="icon" />
            </IconButton>
          </TrackBox>
        );
      })}
    </Wrapper>
  );
}

PlayList.propTypes = {
  musics: PropTypes.array.isRequired,
};
