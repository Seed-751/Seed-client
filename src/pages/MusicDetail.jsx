import React, { useState, useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { useParams, useHistory } from "react-router-dom";
import styled from "styled-components";

import BottomPlayer from "../components/BottomPlayer";
import AlbumInfo from "../components/AlbumInfo";
import PlayList from "../components/PlayList";

import requestMusic from "../api/requestMusic";
import { selectUser } from "../reducers/userSlice";
import { occurError } from "../reducers/errorSlice";

const Container = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  width: 80%;
  margin-top: 3%;
  margin-left: 10%;
`;

export default function MusicDetail() {
  const params = useParams();
  const history = useHistory();
  const musicId = params.music_id;
  const { userInfo } = useSelector(selectUser);
  const dispatch = useDispatch();
  const [album, setAlbum] = useState(null);
  const [currentTrack, setCurrentTrack] = useState(null);

  useEffect(() => {
    async function getMusic() {
      try {
        const { data, message } = await requestMusic(musicId);

        if (data) {
          setCurrentTrack(data.audios[0]);
          return setAlbum(data);
        }

        dispatch(occurError(message));
      } catch (err) {
        dispatch(occurError(err));
      }
    }

    getMusic();
  }, [musicId, dispatch]);

  function handleCurrentTrack(music) {
    setCurrentTrack(music);
  }

  if (!album) {
    return null;
  }

  function handlePaymentPage() {
    history.push({
      pathname: `/payment/${album._Id}/${userInfo._id}`,
      albumInfo: album,
      userInfo,
    });
  }

  return (
    <>
      {album &&
        <Container>
          <AlbumInfo album={album} onClick={handlePaymentPage} />
          <PlayList musics={album.audios} onClick={handleCurrentTrack} />
          <BottomPlayer music={currentTrack} image={album.image} />
        </Container>
      }
    </>
  );
}
