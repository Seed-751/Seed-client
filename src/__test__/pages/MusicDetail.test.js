import React from "react";
import { render } from "@testing-library/react";

import MusicDetail from "../../pages/MusicDetail";
import MockProvider from "../MockProvider";

describe("MusicDetail page", () => {
  afterEach(() => {
    jest.restoreAllMocks();
  });

  it("matches snapshot", () => {
    const { container } = render(
      <MockProvider>
        <MusicDetail />
      </MockProvider>
    );

    expect(container).toMatchSnapshot();
  });
});
