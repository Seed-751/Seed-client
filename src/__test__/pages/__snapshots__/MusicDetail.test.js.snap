// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MusicDetail page matches snapshot 1`] = `<div />`;
